### PR TITLE
Sync contact urns from Identity Store

### PR DESCRIPTION
**Pull request by KaitCrawford**

### CODE REVIEW

**Comment by KaitCrawford** [comment]
@rudigiesler ready for review

**Comment by KaitCrawford** [comment]
@rowanseymour This is for syncing Contact URNs in the Junebug backend. I'm not sure if you're interested in it but I thought I'd let you know it's coming anyway.

File: casepro/backend/junebug.py [hunk 1]
```diff
@@ -109,9 +109,15 @@ def __init__(self, json_data):
...
-        self.addresses = json_data.get('details').get('addresses')
         self.fields = {}
         self.groups = {}
+        addresses = json_data.get('details').get('addresses')
+        self.urns = []
+        for scheme, address in addresses.items():
+            for urn, details in address.items():
+                if 'optedout' in details and details['optedout'] is True:
+                    continue
+                self.urns.append(scheme+":"+urn)
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: I'm wondering if we should take in consideration the `default` detail. Say for the `phone` URN type, if there are multiple addresses with the `phone` type, and one of them has the `default=True` detail, then maybe we don't add the others to the contact, only the ones that are default (and if none are default, then we don't add all like usual)

**Comment by rudigiesler** [comment]
Just the one comment from me, otherwise it looks good 😄

File: casepro/backend/tests/test_junebug.py [hunk 2]
```diff
@@ -165,6 +168,7 @@ def test_pull_contacts_recently_updated(self):
         [contact] = Contact.objects.all()
         self.assertEqual(contact.uuid, "test_id")
         self.assertEqual(contact.name, "test")
+        self.assertEqual(contact.urns, ["msisdn:+1234"])
```
Comment (rowanseymour) [inline casepro/backend/tests/test_junebug.py]: Currently for outgoing messages the scheme for phone numbers is `tel:`

**Comment by KaitCrawford** [comment]
@rudigiesler ready for re-review

File: casepro/backend/junebug.py [hunk 3]
```diff
@@ -109,9 +109,25 @@ def __init__(self, json_data):
...
+                if 'optedout' in details and details['optedout'] is True:
+                    # Skip opted out URNs
+                    continue
+                if 'default' in details and details['default'] is True:
+                    # If a default is set for the scheme then only store the default
+                    scheme_addresses = [urn]
+                    break
+                scheme_addresses.append(urn)
+            for value in scheme_addresses:
+                if scheme == "msisdn":
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: I'm not sure we actually need to do this? `msisdn` and `tel` don't mean anything to the identity store, it's just a convention that we have been using.

File: casepro/backend/junebug.py [hunk 4]
```diff
@@ -109,9 +109,25 @@ def __init__(self, json_data):
...
+                    scheme = "tel"
+                self.urns.append(scheme+":"+value)
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: We either need spaces around the `+`, or we can use `'%s:%s' % (...)` or `.format`.

File: casepro/backend/junebug.py [hunk 5]
```diff
@@ -140,6 +157,10 @@ def update_required(self, local, remote, remote_as_kwargs):
         if {g.uuid for g in local.groups.all()} != {g.uuid for g in remote.groups}:
             return True
 
+        urn_diff = set(local.urns).symmetric_difference(set(remote.urns))
+        if urn_diff != set([]):
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: We can just do an `if urn_diff` or `if bool(urn_diff)` here instead.

**Comment by rudigiesler** [comment]
Just those few comments from me, but looking good 😄

File: casepro/backend/junebug.py (hunk 3 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: @rudigiesler but how is this currently being added then? It's in on the data we send to the Identity Store, is Junebug sending that? It isn't configurable is it?

Comment (rudigiesler) [inline casepro/backend/junebug.py]: Junebug doesn't know anything about address schemes. If the address doesn't exist in the identity store, then when we create it we use the value in the settings file `IDENTITY_ADDRESS_TYPE` (which is currently set to default to `msisdn`, we should probably change it to default to `tel`, but that might mean changing a bunch of tests.)

So it is configurable (at least when casepro is the one creating the identity, and we don't have any other services that create identities that I know of)

**Comment by KaitCrawford** [comment]
@rudigiesler ready for re-review

**Comment by rudigiesler** [comment]
This looks good to me. It seems like your changes broke a test in the base branch. Maybe try a rebase/merge to get up to date, and fix the test. I'm :+1: though otherwise.

**Pull request merged**